Core/fft_mt_r2iq.cpp

```diff
@@ -310,45 +336,76 @@ void * fft_mt_r2iq::r2iqThreadf(r2iqThreadArg *th) {
 		// decimate in frequency plus tuning
 
 		// Caculate the parameters for the first half
-		auto count = std::min(mfft/2, halfFft - _mtunebin);
-		auto source = &th->ADCinFreq[_mtunebin];
+		const auto count = std::min(mfft/2, halfFft - _mtunebin);
+		const auto source = &th->ADCinFreq[_mtunebin];
 
 		// Caculate the parameters for the second half
-		auto start = std::max(0, mfft / 2 - _mtunebin);
-		auto source2 = &th->ADCinFreq[_mtunebin - mfft / 2];
-		auto filter2 = &filter[halfFft - mfft / 2];
-		auto dest = &th->inFreqTmp[mfft / 2];
+		const auto start = std::max(0, mfft / 2 - _mtunebin);
+		const auto source2 = &th->ADCinFreq[_mtunebin - mfft / 2];
+		const auto filter2 = &filter[halfFft - mfft / 2];
+		const auto dest = &th->inFreqTmp[mfft / 2];
 		for (int k = 0; k < fftPerBuf; k++)
 		{
-			// FFT first stage time to frequency, real to complex
-			fftwf_execute_dft_r2c(th->plan_t2f_r2c, th->ADCinTime + (3 * halfFft / 2) * k, th->ADCinFreq);
-
-			for (int m = 0; m < count; m++) // circular shift tune fs/2 first half array
+			// core of fast convolution including filter and decimation
+			//   main part is 'overlap-scrap' (IMHO better name for 'overlap-save'), see
+			//   https://en.wikipedia.org/wiki/Overlap%E2%80%93save_method
 			{
-				th->inFreqTmp[m][0] = (source[m][0] * filter[m][0] +
-										source[m][1] * filter[m][1]);
-
-				th->inFreqTmp[m][1] = (source[m][1] * filter[m][0] -
-										source[m][0] * filter[m][1]);
-				}
-			if (mfft/2 != count)
-					memset(th->inFreqTmp[count], 0, sizeof(float) * 2 * (mfft/2 - count));
+				// FFT first stage: time to frequency, real to complex
+				// 'full' transformation size: 2 * halfFft
+				fftwf_execute_dft_r2c(th->plan_t2f_r2c, th->ADCinTime + (3 * halfFft / 2) * k, th->ADCinFreq);
+				// result now in th->ADCinFreq[]
 
-			for (int m = start; m < mfft / 2; m++) // circular shift tune fs/2 second half array
-			{
-				dest[m][0] = (source2[m][0] * filter2[m][0] +
-								source2[m][1] * filter2[m][1]);
+				// circular shift (mixing in full bins) and low/bandpass filtering (complex multiplication)
+				{
```

Comment:
Can we optimize here if decimate = 0, then we don't need a digit filter?

Reply:
From bandwidth perspective, you could be right .. but, with really unsure about this.
We could try to derive this from formulas .. of overlap-scrap .. or simply try-out!

Reply:
i fear we get a very very low stopband attenuation of 13 dB for the rectangle window ..

Reply:
so 120db here is just fooling ourselves?

Reply:
we are now mixing things
* 120 dB is surely some fooling .. cause the Kaiser window function won't be able to create such a good filter with such few coefficients in relation to fft size **for high decimation**
* kaiser window is applied/integrated into the low/bandpass filter coefficients .. if we optimize for decimation=0 by removing the complex multiplication with filter2[] .. we remove the windowing .. and end up with sort of rectangle

Reply:
> Can we optimize here if decimate = 0, then we don't need a digit filter?

putting the for loops with complex multiplication into an if .. replacing the complex multiplication by simple copy for decimate==0 .. still looks quite good on VHF - except the attenuation is really missing!, which ist sort of integrated in the filter coefficients. thus, it's really necessary to test on HF without the R820T's filters covering problems